File: app/src/main/java/com/fastaccess/provider/timeline/handler/TableHandler.java
package com.fastaccess.provider.timeline.handler;

import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.ColorFilter;
import android.graphics.Paint;
import android.graphics.Paint.Style;
import android.graphics.PixelFormat;
import android.graphics.Typeface;
import android.graphics.drawable.Drawable;
import android.support.annotation.NonNull;
import android.text.Layout.Alignment;
import android.text.Spannable;
import android.text.SpannableStringBuilder;
import android.text.Spanned;
import android.text.StaticLayout;
import android.text.TextPaint;
import android.text.style.AlignmentSpan;
import android.text.style.ImageSpan;

import net.nightwhistler.htmlspanner.TagNodeHandler;

import org.htmlcleaner.TagNode;

import java.util.ArrayList;
import java.util.List;

/**
 * Handles simple HTML tables.
 * <p>
 * Since it renders these tables itself, it needs to know things like font size
 * and text colour to use.
 *
 * @author Alex Kuiper
 */
public class TableHandler extends TagNodeHandler {

    private int tableWidth = 500;
    private Typeface typeFace = Typeface.DEFAULT;
    private float textSize = 28f;
    private int textColor = Color.BLACK;
    private static final int PADDING = 20;

    @Override public boolean rendersContent() {
        return true;
    }

    @Override public void handleTagNode(TagNode node, SpannableStringBuilder builder, int start, int end) {
        Table table = getTable(node);
        for (int i = 0; i < table.getRows().size(); i++) {
            List<Spanned> row = table.getRows().get(i);
            builder.append("\uFFFC");
            TableRowDrawable drawable = new TableRowDrawable(row, table.isDrawBorder());
            drawable.setBounds(0, 0, drawable.getIntrinsicWidth(),
                    drawable.getIntrinsicHeight());
            builder.setSpan(new ImageSpan(drawable), start + i, builder.length(), 33);

        }
        builder.append("\uFFFC");
        Drawable drawable = new TableRowDrawable(new ArrayList<Spanned>(), table.isDrawBorder());
        drawable.setBounds(0, 0, tableWidth, 1);
        builder.setSpan(new ImageSpan(drawable), builder.length() - 1, builder.length(),
                Spannable.SPAN_EXCLUSIVE_EXCLUSIVE);
        builder.setSpan((AlignmentSpan) () -> Alignment.ALIGN_CENTER, start, builder.length(), Spannable.SPAN_EXCLUSIVE_EXCLUSIVE);
        builder.append("\n");
    }

    public void setTableWidth(int tableWidth) {
        this.tableWidth = tableWidth;
    }

    public void setTextColor(int textColor) {
        this.textColor = textColor;
    }

    private void readNode(Object node, Table table) {
        if (node instanceof TagNode) {
            TagNode tagNode = (TagNode) node;
            if (tagNode.getName().equals("td") || tagNode.getName().equals("th")) {
                Spanned result = this.getSpanner().fromTagNode(tagNode);
                table.addCell(result);
                return;
            }
            if (tagNode.getName().equals("tr")) {
                table.addRow();
            }
            for (Object child : tagNode.getChildTags()) {
                readNode(child, table);
            }
        }

    }

    private Table getTable(TagNode node) {

        String border = node.getAttributeByName("border");

        boolean drawBorder = !"0".equals(border);

        Table result = new Table(drawBorder);

        readNode(node, result);

        return result;
    }

    private TextPaint getTextPaint() {
        TextPaint textPaint = new TextPaint();
        textPaint.setColor(this.textColor);
        textPaint.linkColor = this.textColor;
        textPaint.setAntiAlias(true);
        textPaint.setTextSize(this.textSize);
        textPaint.setTypeface(this.typeFace);

        return textPaint;
    }

    private int calculateRowHeight(List<Spanned> row) {

        if (row.size() == 0) {
            return 0;
        }

        TextPaint textPaint = getTextPaint();

        int columnWidth = tableWidth / row.size();

        int rowHeight = 0;

        for (Spanned cell : row) {

            StaticLayout layout = new StaticLayout(cell, textPaint, columnWidth
                    - 2 * PADDING, Alignment.ALIGN_NORMAL, 1.5f, 0.5f, true);

            if (layout.getHeight() > rowHeight) {
                rowHeight = layout.getHeight();
            }
        }

        return rowHeight;
    }

    private class TableRowDrawable extends Drawable {

        private List<Spanned> tableRow;

        private int rowHeight;
        private boolean paintBorder;

        TableRowDrawable(List<Spanned> tableRow, boolean paintBorder) {
            this.tableRow = tableRow;
            this.rowHeight = calculateRowHeight(tableRow);
            this.paintBorder = paintBorder;
        }

        @Override public void draw(@NonNull Canvas canvas) {
            Paint paint = new Paint();
            paint.setColor(textColor);
            paint.setStyle(Style.STROKE);

            int numberOfColumns = tableRow.size();

            if (numberOfColumns == 0) {
                return;
            }

            int columnWidth = tableWidth / numberOfColumns;

            int offset;

            for (int i = 0; i < numberOfColumns; i++) {

                offset = i * columnWidth;

                if (paintBorder) {
                    // The rect is open at the bottom, so there's a single line
                    // between rows.
                    canvas.drawRect(offset, 0, offset + columnWidth, rowHeight, paint);
                }

                StaticLayout layout = new StaticLayout(tableRow.get(i),
                        getTextPaint(), (columnWidth - 2 * PADDING),
                        Alignment.ALIGN_NORMAL, 1.5f, 0.5f, true);

                canvas.translate(offset + PADDING, 0);
                layout.draw(canvas);
                canvas.translate(-1 * (offset + PADDING), 0);

            }
        }

        @Override
        public int getIntrinsicHeight() {
            return rowHeight;
        }

        @Override
        public int getIntrinsicWidth() {
            return tableWidth;
        }

        @Override
        public int getOpacity() {
            return PixelFormat.OPAQUE;
        }

        @Override
        public void setAlpha(int alpha) {

        }

        @Override
        public void setColorFilter(ColorFilter cf) {

        }
    }

    private class Table {

        private boolean drawBorder;
        private List<List<Spanned>> content = new ArrayList<>();

        private Table(boolean drawBorder) {
            this.drawBorder = drawBorder;
        }

        boolean isDrawBorder() {
            return drawBorder;
        }

        void addRow() {
            content.add(new ArrayList<>());
        }

        List<Spanned> getBottomRow() {
            return content.get(content.size() - 1);
        }

        List<List<Spanned>> getRows() {
            return content;
        }

        void addCell(Spanned text) {
            if (content.isEmpty()) {
                throw new IllegalStateException("No rows added yet");
            }

            getBottomRow().add(text);
        }
    }

}

File: markdown/src/main/java/com/fastaccess/markdown/spans/TableHandler.kt
package com.fastaccess.markdown.spans

import android.graphics.*
import android.graphics.Paint.Style
import android.graphics.drawable.Drawable
import android.text.*
import android.text.Layout.Alignment
import android.text.style.AlignmentSpan
import android.text.style.ImageSpan
import net.nightwhistler.htmlspanner.SpanStack
import net.nightwhistler.htmlspanner.TagNodeHandler
import org.htmlcleaner.TagNode
import java.util.*

/**
 * Handles simple HTML tables.
 *
 *
 * Since it renders these tables itself, it needs to know things like font size
 * and text colour to use.
 *
 * @author Alex Kuiper
 */
class TableHandler(
        private val textColor: Int = Color.BLACK,
        private val tableWidth: Int = 500
) : TagNodeHandler() {
    private val textPaint: TextPaint
        get() {
            val textPaint = TextPaint()
            textPaint.color = this.textColor
            textPaint.linkColor = this.textColor
            textPaint.isAntiAlias = true
            textPaint.textSize = 28f
            textPaint.typeface = Typeface.DEFAULT
            return textPaint
        }

    override fun rendersContent() = true

    override fun handleTagNode(node: TagNode?, builder: SpannableStringBuilder?, start: Int, end: Int, spanStack: SpanStack?) {
        builder?.let {
            val table = getTable(node)
            for (i in 0 until table.rows.size) {
                val row = table.rows[i]
                builder.append("\uFFFC")
                val drawable = TableRowDrawable(row, table.isDrawBorder)
                drawable.setBounds(0, 0, drawable.intrinsicWidth,
                        drawable.intrinsicHeight)
                builder.setSpan(ImageSpan(drawable), start + i, builder.length, 33)

            }
            builder.append("\uFFFC")
            val drawable = TableRowDrawable(ArrayList(), table.isDrawBorder)
            drawable.setBounds(0, 0, tableWidth, 1)
            builder.setSpan(ImageSpan(drawable), builder.length - 1, builder.length,
                    Spannable.SPAN_EXCLUSIVE_EXCLUSIVE)
            builder.setSpan({ Alignment.ALIGN_CENTER } as AlignmentSpan, start, builder.length, Spannable.SPAN_EXCLUSIVE_EXCLUSIVE)
            builder.append("\n")
        }
    }

    private fun readNode(node: Any?, table: Table) {
        if (node is TagNode) {
            if (node.name == "td" || node.name == "th") {
                val result = this.spanner.fromTagNode(node, null)
                table.addCell(result)
                return
            }
            if (node.name == "tr") {
                table.addRow()
            }
            for (child in node.childTags) {
                readNode(child, table)
            }
        }

    }

    private fun getTable(node: TagNode?): Table {
        val border = node?.getAttributeByName("border")
        val drawBorder = "0" != border
        val result = Table(drawBorder)
        readNode(node, result)
        return result
    }

    private fun calculateRowHeight(row: List<Spanned>): Int {

        if (row.isEmpty()) {
            return 0
        }

        val textPaint = textPaint

        val columnWidth = tableWidth / row.size

        var rowHeight = 0

        for (cell in row) {

            val layout = StaticLayout(cell, textPaint, columnWidth - 2 * PADDING, Alignment.ALIGN_NORMAL, 1.5f, 0.5f, true)

            if (layout.height > rowHeight) {
                rowHeight = layout.height
            }
        }

        return rowHeight
    }

    private inner class TableRowDrawable internal constructor(private val tableRow: List<Spanned>, private val paintBorder: Boolean) : Drawable() {

        private val rowHeight: Int

        init {
            this.rowHeight = calculateRowHeight(tableRow)
        }

        override fun draw(canvas: Canvas) {
            val paint = Paint()
            paint.color = textColor
            paint.style = Style.STROKE

            val numberOfColumns = tableRow.size

            if (numberOfColumns == 0) {
                return
            }

            val columnWidth = tableWidth / numberOfColumns

            var offset: Int

            for (i in 0 until numberOfColumns) {

                offset = i * columnWidth

                if (paintBorder) {
                    // The rect is open at the bottom, so there's a single line
                    // between rows.
                    canvas.drawRect(offset.toFloat(), 0f, (offset + columnWidth).toFloat(), rowHeight.toFloat(), paint)
                }

                val layout = StaticLayout(tableRow[i],
                        textPaint, columnWidth - 2 * PADDING,
                        Alignment.ALIGN_NORMAL, 1.5f, 0.5f, true)

                canvas.translate((offset + PADDING).toFloat(), 0f)
                layout.draw(canvas)
                canvas.translate((-1 * (offset + PADDING)).toFloat(), 0f)

            }
        }

        override fun getIntrinsicHeight(): Int {
            return rowHeight
        }

        override fun getIntrinsicWidth(): Int {
            return tableWidth
        }

        override fun getOpacity(): Int {
            return PixelFormat.OPAQUE
        }

        override fun setAlpha(alpha: Int) {

        }

        override fun setColorFilter(cf: ColorFilter?) {

        }
    }

    private inner class Table constructor(internal val isDrawBorder: Boolean) {
        private val content = ArrayList<List<Spanned>>()

        internal val bottomRow: MutableList<Spanned>
            get() = content[content.size - 1] as MutableList<Spanned>

        internal val rows: List<List<Spanned>>
            get() = content

        internal fun addRow() {
            content.add(ArrayList())
        }

        internal fun addCell(text: Spanned) {
            if (content.isEmpty()) {
                throw IllegalStateException("No rows added yet")
            }

            bottomRow.add(text)
        }
    }

    companion object {
        private const val PADDING = 20
    }

}